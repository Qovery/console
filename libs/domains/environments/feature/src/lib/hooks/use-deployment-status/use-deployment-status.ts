import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseDeploymentStatusProps {
  environmentId?: string
}

export function useDeploymentStatus({ environmentId }: UseDeploymentStatusProps) {
  return useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.environments.deploymentStatus(environmentId!!),
    enabled: Boolean(environmentId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })
}

export default useDeploymentStatus
