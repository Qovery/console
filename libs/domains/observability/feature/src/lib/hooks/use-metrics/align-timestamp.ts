// Helpers for alignment (timestamps in seconds)
// Needed to avoid issues with Prometheus when the time range is not aligned with the step interval
const ALIGN_SEC = 30
export const alignStartSec = (ts?: string) =>
  ts == null ? undefined : Math.floor(Number(ts) / ALIGN_SEC) * ALIGN_SEC + ''
export const alignEndSec = (ts?: string) =>
  ts == null ? undefined : Math.ceil(Number(ts) / ALIGN_SEC) * ALIGN_SEC + ''

const THANOS_RAW_RETENTION = 15 * 24 * 60 * 60 * 1000 // 15 days
const THANOS_5M_RETENTION = 30 * 24 * 60 * 60 * 1000 // 30 days

export const resolutionByRetention = (startTimestamp: string) => {
  const startMs = Number(startTimestamp) * 1000
  const nowMs = Date.now()
  const ageMs = Math.max(0, nowMs - startMs)

  let byAge: '0s' | '5m' | '1h'
  if (ageMs > THANOS_RAW_RETENTION) {
    if (ageMs > THANOS_5M_RETENTION) {
      byAge = '1h'
    } else {
      byAge = '5m'
    }
  } else {
    byAge = '0s'
  }

  return byAge
}
