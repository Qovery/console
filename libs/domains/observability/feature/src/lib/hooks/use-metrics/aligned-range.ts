const DURATION_TO_MINUTES: Record<string, number> = {
  '1m': 1,
  '5m': 5,
  '15m': 15,
  '30m': 30,
  '1h': 60,
  '3h': 180,
  '6h': 360,
  '12h': 720,
  '24h': 1440,
  '2d': 2880,
  '7d': 10080,
  '14d': 20160,
  '21d': 30240,
  '28d': 40320,
}

export const alignedRangeInMinutes = (startTimestamp?: string, endTimestamp?: string) => {
  if (!startTimestamp || !endTimestamp) {
    return 'na'
  }

  const start = parseInt(startTimestamp, 10)
  const end = parseInt(endTimestamp, 10)

  if (isNaN(start) || isNaN(end)) {
    return 'na'
  }

  const diffSeconds = end - start
  const diffMinutes = diffSeconds / 60

  const supportedMinutes = Object.values(DURATION_TO_MINUTES).sort((a, b) => a - b)

  let selectedRange = supportedMinutes[0]

  for (const range of supportedMinutes) {
    if (diffMinutes >= range) {
      selectedRange = range
    } else {
      break
    }
  }

  return selectedRange.toString().padStart(6, '0')
}
