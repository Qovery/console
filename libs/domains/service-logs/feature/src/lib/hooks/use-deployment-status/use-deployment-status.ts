import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseDeploymentStatusProps {
  environmentId?: string
  serviceId?: string
}

export function useDeploymentStatus({ environmentId, serviceId }: UseDeploymentStatusProps) {
  return useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.services.deploymentStatus(environmentId!!, serviceId!!),
    enabled: Boolean(environmentId) && Boolean(serviceId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })
}

export default useDeploymentStatus
