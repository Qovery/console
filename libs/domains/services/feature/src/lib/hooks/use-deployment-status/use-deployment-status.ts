import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'
import useListStatuses from '../use-list-statuses/use-list-statuses'

export interface UseDeploymentStatusProps {
  environmentId?: string
  serviceId?: string
}

export function useDeploymentStatus({ environmentId, serviceId }: UseDeploymentStatusProps) {
  const webSocketResult = useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.services.deploymentStatus(environmentId!!, serviceId!!),
    enabled: Boolean(environmentId) && Boolean(serviceId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })

  // NOTE: Deployment status by WS isn't enough because WS can something return no message.
  // We need to fallback on service status, unfortunately we have to play with endpoints to get
  // a service status without knowing its type (eg. application, container ...)
  const { data: environmentStatus, ...rest } = useListStatuses({ environmentId })
  const serviceStatuses = [
    ...(environmentStatus?.applications ?? []),
    ...(environmentStatus?.containers ?? []),
    ...(environmentStatus?.databases ?? []),
    ...(environmentStatus?.jobs ?? []),
  ]
  return webSocketResult.data
    ? webSocketResult
    : {
        data: serviceStatuses.find(({ id }) => id === serviceId),
        ...rest,
      }
}

export default useDeploymentStatus
