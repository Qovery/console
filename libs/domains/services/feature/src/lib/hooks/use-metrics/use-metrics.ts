import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseMetricsProps {
  environmentId?: string
  serviceId?: string
}

export function useMetrics({ environmentId, serviceId }: UseMetricsProps) {
  return useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.services.metrics(environmentId!!, serviceId!!),
    select(data) {
      if (!data) {
        return data
      }
      return data.sort((a, b) => a.pod_name.localeCompare(b.pod_name))
    },
    enabled: Boolean(environmentId) && Boolean(serviceId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })
}

export default useMetrics
