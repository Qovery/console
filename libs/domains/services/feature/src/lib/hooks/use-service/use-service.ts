import { type UseQueryResult, useQuery } from '@tanstack/react-query'
import {
  type AnyService,
  type Application,
  type ApplicationType,
  type Container,
  type ContainerType,
  type Database,
  type DatabaseType,
  type Helm,
  type HelmType,
  type Job,
  type JobType,
  type ServiceType,
} from '@qovery/domains/services/data-access'
import { queries } from '@qovery/state/util-queries'
import { useServiceType } from '../use-service-type/use-service-type'

export type UseServiceProps =
  | {
      environmentId?: string
      serviceId?: string
    }
  | {
      serviceId?: string
      serviceType: ServiceType
    }

/**
 * Here we use TS function overload to allow us to call the hook in 2 different ways.
 * The first one to get a service which you don't know the type of ahead.
 * ```ts
 * const { data: service } = useService({ environmentId: '123', serviceId: '456' })
 * ```
 * The seconde one to get a service which you already know the type of.
 * ```ts
 * const { data: database } = useService({ serviceId: '456', serviceType: 'DATABASE' })
 * ```
 *
 * https://dev.to/zirkelc/how-to-return-different-types-from-functions-in-typescript-2a2h
 * https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads
 **/
export function useService(props: { environmentId?: string; serviceId?: string }): UseQueryResult<AnyService>
export function useService<
  T extends ServiceType,
  R = T extends ApplicationType
    ? Application
    : T extends ContainerType
      ? Container
      : T extends DatabaseType
        ? Database
        : T extends JobType
          ? Job
          : T extends 'CRON_JOB'
            ? Job
            : T extends 'LIFECYCLE_JOB'
              ? Job
              : T extends HelmType
                ? Helm
                : never,
>(props: { serviceId: string; serviceType: T }): UseQueryResult<R>
export function useService({ serviceId, ...props }: UseServiceProps) {
  const { data: serviceType } = useServiceType({
    environmentId: 'environmentId' in props ? props.environmentId : undefined,
    serviceId,
  })

  return useQuery({
    ...queries.services.details({
      serviceId: serviceId!,
      serviceType: 'serviceType' in props ? props.serviceType : serviceType!,
    }),
    enabled: Boolean(serviceId) && Boolean(serviceType),
  })
}

export default useService
