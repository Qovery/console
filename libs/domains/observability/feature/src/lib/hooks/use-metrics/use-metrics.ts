import { useQuery } from '@tanstack/react-query'
import { useEffect, useMemo, useRef, useState } from 'react'
import { observability } from '@qovery/domains/observability/data-access'
import { useServiceOverviewContext } from '../../service-overview/util-filter/service-overview-context'
import { type TimeRangeOption } from '../../service-overview/util-filter/time-range'
import { alignEndSec, alignStartSec } from './align-timestamp'
import { alignedRangeInMinutes } from './grafana-util'

export interface MetricData {
  metric: {
    container: string
    cpu: string
    endpoint: string
    id: string
    image: string
    instance: string
    job: string
    metrics_path: string
    name: string
    namespace: string
    node: string
    pod: string
    prometheus: string
    service: string
    reason?: string
  }
  values: [number, string][]
}

interface UseMetricsProps {
  clusterId: string
  query: string
  startTimestamp?: string
  endTimestamp?: string
  timeRange?: TimeRangeOption
  isLiveUpdateEnabled?: boolean
  overriddenResolution?: string
  overriddenMaxPoints?: number
  boardShortName: 'service_overview'
  metricShortName: string
}

function useLiveUpdateSetting(): boolean {
  const context = useServiceOverviewContext()
  // Pause live updates when charts are zoomed or when the date picker is open
  return context.isLiveUpdateEnabled && !context.isAnyChartZoomed && !context.isDatePickerOpen
}

// Simple wrapper that automatically applies live update toggle from context
export function useMetrics({
  clusterId,
  query,
  startTimestamp,
  endTimestamp,
  timeRange,
  isLiveUpdateEnabled: overrideLiveUpdate,
  overriddenResolution,
  overriddenMaxPoints,
  boardShortName,
  metricShortName,
}: UseMetricsProps) {
  // Get context and live update setting, but allow override
  const context = useServiceOverviewContext()
  const contextLiveUpdate = useLiveUpdateSetting()
  const finalLiveUpdateEnabled = overrideLiveUpdate ?? contextLiveUpdate

  const alignedStart = alignStartSec(startTimestamp)
  const alignedEnd = alignEndSec(endTimestamp)
  const alignedRange = alignedRangeInMinutes(alignedStart, alignedEnd)

  const step = useMemo(() => {
    // Actual scrape_interval = 30s
    // https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration
    // https://qovery.slack.com/archives/C02NQ0LC8M9/p1753804730631609?thread_ts=1753803872.246399&cid=C02NQ0LC8M9
    if (timeRange === '5m') return '30000ms' // 30 seconds (match actual scrape_interval)
    if (timeRange === '15m') return '60000ms' // 1 minute for longer ranges (2x scrape_interval)
    if (timeRange === '30m') return '120000ms' // 2 minutes for longer ranges (4x scrape_interval)
    if (alignedStart && alignedEnd) {
      return calculateDynamicRange(alignedStart, alignedEnd, overriddenMaxPoints)
    }
    return '30000ms' // Default: 30 seconds (match actual scrape_interval)
  }, [timeRange, alignedStart, alignedEnd, overriddenMaxPoints])

  const maxSourceResolution = useMemo(() => {
    if (overriddenResolution !== undefined) {
      return overriddenResolution ?? ('0s' as const)
    }
    if (!alignedStart || !alignedEnd) return '0s' as const
    const safeStep = step ?? '1m'
    return calculateMaxSourceResolution(alignedStart, alignedEnd, safeStep)
  }, [alignedStart, alignedEnd, step, overriddenResolution])

  const queryResult = useQuery({
    ...observability.metrics({
      clusterId,
      query,
      queryRange: 'query_range',
      startTimestamp: alignedStart,
      endTimestamp: alignedEnd,
      timeRange,
      step,
      maxSourceResolution,
      // These params are used to generate charts in Grafana
      boardShortName,
      metricShortName,
      traceId: 'na',
      alignedRange,
    }),
    keepPreviousData: true,
    refetchInterval: finalLiveUpdateEnabled ? 30_000 : false, // Refetch every 30 seconds only if live update is enabled
    refetchIntervalInBackground: finalLiveUpdateEnabled,
    refetchOnWindowFocus: false,
    retry: 3,
  })

  // Custom isLoading: true on first load and when timeRange changes, false on live refetch
  const lastTimeRange = useRef<TimeRangeOption | undefined>(timeRange)
  const [isLoadingCustom, setIsLoadingCustom] = useState(queryResult.isLoading)

  useEffect(() => {
    if (timeRange !== lastTimeRange.current) {
      setIsLoadingCustom(true)
      lastTimeRange.current = timeRange
    }
  }, [timeRange])

  useEffect(() => {
    if (!queryResult.isFetching) {
      setIsLoadingCustom(false)
    }
  }, [queryResult.isFetching])

  // isRefreshing: true when fetching data but not on initial load or time range changes
  const isRefreshing = queryResult.isFetching && !isLoadingCustom

  // Track refreshing state at global level through context
  const previousIsRefreshing = useRef(false)
  useEffect(() => {
    if (context && previousIsRefreshing.current !== isRefreshing) {
      // Increment counter when starting to refresh, decrement when stopping
      context.setIsAnyChartRefreshing(isRefreshing)
      previousIsRefreshing.current = isRefreshing
    }
  }, [context, isRefreshing])

  return {
    ...queryResult,
    isLoading: isLoadingCustom,
    isRefreshing,
    stepInSecond: promDurationToMs(step) / 1000,
  }
}

// Grafana-like auto rate interval
export function calculateRateInterval(startTimestamp: string, endTimestamp: string): string {
  const startMs = Number(startTimestamp) * 1000
  const endMs = Number(endTimestamp) * 1000
  const durationMs = endMs - startMs

  // Defaults
  const scrapeIntervalMs = 15_000 // 15s Prom scrape interval
  const targetPoints = 200

  // Allowed rate windows (Grafana-like buckets)
  const allowedStepsMs = [
    60_000, // 1m
    5 * 60_000, // 5m
    15 * 60_000, // 15m
    30 * 60_000, // 30m
    60 * 60_000, // 1h
    2 * 60 * 60_000, // 2h
    6 * 60 * 60_000, // 6h
    12 * 60 * 60_000, // 12h
  ] as const

  // Minimum = 4× scrape interval (avoid jitter)
  const minWindowMs = Math.max(4 * scrapeIntervalMs, allowedStepsMs[0])

  // Ideal = range ÷ targetPoints
  const desiredMs = Math.max(minWindowMs, Math.ceil(durationMs / targetPoints))

  // Pick the closest allowed step >= desired
  const pickedMs = allowedStepsMs.find((s) => s >= desiredMs) ?? allowedStepsMs[allowedStepsMs.length - 1]

  function msToPromDuration(ms: number): string {
    if (ms % (60 * 60 * 1000) === 0) return `${ms / (60 * 60 * 1000)}h`
    if (ms % (60 * 1000) === 0) return `${ms / (60 * 1000)}m`
    if (ms % 1000 === 0) return `${ms / 1000}s`
    return `${ms}ms`
  }

  return msToPromDuration(pickedMs)
}

/**
 * Calculate a Prometheus/Thanos query step.
 *
 * Same API as before:
 *   - startTimestamp, endTimestamp in seconds (string)
 *   - offsetMultiplier in 30s units (optional)
 *   - overriddenMaxPoints = target number of points (default 150)
 *
 * Returns: "<N>ms"
 */
export function calculateDynamicRange(
  startTimestamp: string,
  endTimestamp: string,
  offsetMultiplier = 0,
  overriddenMaxPoints = 150
): string {
  const startMs = Number(startTimestamp) * 1000
  const endMs = Number(endTimestamp) * 1000
  const durationMs = Math.max(0, endMs - startMs)

  // Allowed, quantized step values (ms)
  const allowedStepsMs = [
    30_000, // 30s
    60_000, // 1m
    120_000, // 2m
    300_000, // 5m
    600_000, // 10m
    900_000, // 15m
    1_800_000, // 30m
    3_600_000, // 1h
    7_200_000, // 2h
    10_800_000, // 3h
    21_600_000, // 6h
  ] as const

  // Minimal step based on range / maxPoints
  const rawStep = Math.ceil(durationMs / Math.max(1, overriddenMaxPoints))

  // Snap to grid upwards
  const snapped = allowedStepsMs.find((s) => s >= rawStep) ?? allowedStepsMs[allowedStepsMs.length - 1]

  // Apply jitter: each unit = 30s
  const jitter = offsetMultiplier * 30_000

  // Clamp jitter to ±20% of snapped
  const maxJitter = snapped * 0.2
  const clamped = Math.max(-maxJitter, Math.min(jitter, maxJitter))

  // Final step: keep on grid by snapping up again
  const withJitter = snapped + clamped
  const finalStep = allowedStepsMs.find((s) => s >= withJitter) ?? allowedStepsMs[allowedStepsMs.length - 1]

  // console.log('PGPG calculateDynamicRange ', overriddenMaxPoints, finalStep)
  return `${finalStep}ms`
}

/**
 * Calculate Thanos max_source_resolution based on step and range.
 *
 * Same API as before:
 *   - startTimestamp, endTimestamp in seconds (string)
 *   - renderedStep string ("60000ms", "5m", etc.)
 *
 * Returns one of: "0s" | "5m" | "1h"
 */
export function calculateMaxSourceResolution(
  startTimestamp: string,
  endTimestamp: string,
  renderedStep: string
): '0s' | '5m' | '1h' {
  const stepMs = promDurationToMs(renderedStep)
  const startMs = Number(startTimestamp) * 1000
  const endMs = Number(endTimestamp) * 1000
  const rangeMs = Math.max(0, endMs - startMs)

  // By step (hysteresis thresholds)
  const RAW_BOUNDARY = 150_000 // 2m30s
  const ONE_H_BOUNDARY = 1_800_000 // 30m

  let byStep: '0s' | '5m' | '1h'
  if (stepMs < RAW_BOUNDARY) byStep = '0s'
  else if (stepMs < ONE_H_BOUNDARY) byStep = '5m'
  else byStep = '1h'

  // By range
  const TWELVE_H = 12 * 3_600_000
  const SEVEN_D = 7 * 24 * 3_600_000

  let byRange: '0s' | '5m' | '1h'
  if (rangeMs < TWELVE_H) byRange = '0s'
  else if (rangeMs <= SEVEN_D) byRange = '5m'
  else byRange = '1h'

  // Pick the coarser
  const order = ['0s', '5m', '1h'] as const
  return order[Math.max(order.indexOf(byStep), order.indexOf(byRange))]
}

// Parse duration string into ms
function promDurationToMs(d: string): number {
  const m = d.match(/^(\d+)(ms|s|m|h|d)$/)
  if (!m) return 60_000
  const n = Number(m[1])
  switch (m[2]) {
    case 'ms':
      return n
    case 's':
      return n * 1000
    case 'm':
      return n * 60_000
    case 'h':
      return n * 3_600_000
    case 'd':
      return n * 86_400_000
    default:
      return 60_000
  }
}
