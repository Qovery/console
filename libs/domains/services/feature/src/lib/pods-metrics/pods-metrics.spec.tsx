import { ResourceStatusDto, UnitDto } from 'qovery-ws-typescript-axios'
import { ServiceTypeEnum } from '@qovery/shared/enums'
import { renderWithProviders } from '@qovery/shared/util-tests'
import * as useDeploymentStatusImport from '../hooks/use-deployment-status/use-deployment-status'
import * as useMetricsImport from '../hooks/use-metrics/use-metrics'
import * as useServiceImport from '../hooks/use-service/use-service'
import PodsMetrics from './pods-metrics'

describe('PodsMetrics', () => {
  it('should match snapshot with data', () => {
    jest.spyOn(useMetricsImport, 'useMetrics').mockReturnValue({
      data: [
        {
          cpu: {
            current: 12,
            current_percent: 5,
            limit: 80,
            status: ResourceStatusDto.OK,
            unit: UnitDto.M_CPU,
          },
          memory: {
            current: 30,
            current_percent: 20,
            limit: 80,
            status: ResourceStatusDto.OK,
            unit: UnitDto.MI_B,
          },
          pod_name: 'foobar',
          storages: [],
        },
      ],
      isLoading: false,
      error: {},
      isError: false,
    })
    jest.spyOn(useServiceImport, 'useService').mockReturnValue({
      data: {
        id: '1',
        serviceType: ServiceTypeEnum.APPLICATION,
        created_at: '1696923386000',
        healthchecks: {},
      },
      isLoading: false,
      error: {},
      isError: false,
    })
    const { container } = renderWithProviders(<PodsMetrics environmentId="1" serviceId="1" />)
    expect(container).toMatchSnapshot()
  })
  it('should match snapshot with empty data', () => {
    jest.spyOn(useMetricsImport, 'useMetrics').mockReturnValue({
      data: [],
      isLoading: false,
      error: {},
      isError: false,
    })
    jest.spyOn(useServiceImport, 'useService').mockReturnValue({
      data: {
        id: '1',
        serviceType: ServiceTypeEnum.APPLICATION,
        created_at: '1696923386000',
        healthchecks: {},
      },
      isLoading: false,
      error: {},
      isError: false,
    })
    jest.spyOn(useDeploymentStatusImport, 'useDeploymentStatus').mockReturnValue({
      data: {
        id: '1',
        service_deployment_status: 'UP_TO_DATE',
        state: 'READY',
      },
      isLoading: false,
      error: {},
      isError: false,
    })
    const { container } = renderWithProviders(<PodsMetrics environmentId="1" serviceId="1" />)
    expect(container).toMatchSnapshot()
  })
  it('should match snapshot with null data', () => {
    jest.spyOn(useMetricsImport, 'useMetrics').mockReturnValue({
      data: null,
      isLoading: false,
      error: {},
      isError: false,
    })
    jest.spyOn(useServiceImport, 'useService').mockReturnValue({
      data: {
        id: '1',
        serviceType: ServiceTypeEnum.APPLICATION,
        created_at: '1696923386000',
        healthchecks: {},
      },
      isLoading: false,
      error: {},
      isError: false,
    })
    jest.spyOn(useDeploymentStatusImport, 'useDeploymentStatus').mockReturnValue({
      data: {
        id: '1',
        service_deployment_status: 'UP_TO_DATE',
        state: 'READY',
      },
      isLoading: false,
      error: {},
      isError: false,
    })
    const { container } = renderWithProviders(<PodsMetrics environmentId="1" serviceId="1" />)
    expect(container).toMatchSnapshot()
  })
})
