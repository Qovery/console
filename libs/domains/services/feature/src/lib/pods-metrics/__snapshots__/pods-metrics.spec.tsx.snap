// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodsMetrics should match snapshot with data 1`] = `
<div>
  <div
    class="overflow-x-scroll rounded border xl:overflow-hidden"
  >
    <table
      class="min-w-full divide-y divide-neutral-200 text-neutral-400 w-full overflow-y-scroll text-xs xl:overflow-auto"
    >
      <thead>
        <tr>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Pods (1)
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-down text-xs"
              />
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Status
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Version
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Memory
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              vCPU
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Storage
            </button>
          </th>
          <th
            class="h-11 px-4 text-left font-medium first:w-1/4 first:pl-[52px]"
          >
            <button
              class="flex items-center gap-1 cursor-pointer select-none"
              type="button"
            >
              Age
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        class="divide-y divide-neutral-200"
      >
        <tr
          class="hover:bg-neutral-100"
        >
          <td
            class="h-14 px-4 py-0 text-nowrap"
          >
            <button
              class="text-md pointer inline-flex h-14 w-9 items-center justify-start text-neutral-350"
              type="button"
            >
              <span
                class="icon-solid-chevron-down shrink-0 pl-1"
                role="img"
              />
            </button>
            <button
              class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-xs h-7 px-2 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400 truncate"
            >
              foobar
            </button>
          </td>
          <td
            class="h-14 px-4 py-0"
          >
            <span
              class="inline-flex items-center shrink-0 font-medium text-xs h-6 px-2 border rounded-full border-neutral-250 text-neutral-350 dark:border-neutral-350 dark:text-neutral-300 gap-2 capitalize"
            >
              <div
                class=""
                data-state="closed"
              >
                <svg
                  class="text-neutral-300"
                  fill="none"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    clip-path="url(#clip0_5911_150)"
                  >
                    <path
                      d="M21.75 12a9.75 9.75 0 10-19.5 0 9.75 9.75 0 0019.5 0zM0 12a12 12 0 1124 0 12 12 0 01-24 0zm7.875-3.375A2.627 2.627 0 0110.5 6h2.667a2.958 2.958 0 012.958 2.958c0 1.06-.567 2.039-1.486 2.569l-1.514.867V13.875h-2.25v-2.784l.567-.324 2.077-1.19a.709.709 0 00-.352-1.322H10.5a.376.376 0 00-.375.375v.304h-2.25v-.309zm3 7.125h2.25V18h-2.25v-2.25z"
                      fill="currentColor"
                    />
                  </g>
                  <defs>
                    <clippath
                      id="clip0_5911_150"
                    >
                      <path
                        d="M0 0H24V24H0z"
                        fill="#fff"
                      />
                    </clippath>
                  </defs>
                </svg>
              </div>
              <span
                class="text-neutral-400"
              >
                Unknown
              </span>
            </span>
          </td>
          <td
            class="h-14 px-4 py-0"
          />
          <td
            class="h-14 px-4 py-0"
          >
            30 MiB (20%)
          </td>
          <td
            class="h-14 px-4 py-0"
          >
            12 mCPU (5%)
          </td>
          <td
            class="h-14 px-4 py-0"
          >
            -
          </td>
          <td
            class="h-14 px-4 py-0"
          >
            <i
              aria-hidden="true"
              class="fa-regular fa-circle-question text-sm text-neutral-300"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`PodsMetrics should match snapshot with empty data 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-1 border border-neutral-200 bg-neutral-100 py-10 text-sm text-neutral-350"
  >
    <span
      class="icon-solid-play shrink-0 text-md text-neutral-300"
      role="img"
    />
    <span
      class="font-medium"
    >
      Application is not running
    </span>
    <span>
      Deploy the application first
    </span>
  </div>
</div>
`;

exports[`PodsMetrics should match snapshot with null data 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-1 border border-neutral-200 bg-neutral-100 py-10 text-sm text-neutral-350"
  >
    <span
      class="icon-solid-play shrink-0 text-md text-neutral-300"
      role="img"
    />
    <span
      class="font-medium"
    >
      Application is not running
    </span>
    <span>
      Deploy the application first
    </span>
  </div>
</div>
`;
